### PR TITLE
80 increase cht sync throughput

### LINKED ISSUE #80: Increase cht-sync throughput
Cht-sync is currently running aginst the https://brac-clone-for-phil.dev.medicmobile.org instance but syncing of documents is currently very slow. When I checked a few hours ago it had a backlog of 41,168,339 documents to be synced (see screenshot below).

Image

### PR DESCRIPTION
**Pull request by njuguna-n**
Closes #80

### CODE REVIEW

**Comment by njuguna-n** [comment]
@witash @lorerod please have a preliminary look as I fix the failing tests and Sonar issues.

File: redis-worker/src/config.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+dotenv.config();
+
+const config = {
+  redis: {
+    host: process.env.REDIS_HOST,
+    port: process.env.REDIS_PORT,
+    key: process.env.REDIS_KEY,
+  },
+  postgrest: {
+    url: `http://${process.env.POSTGREST_ENDPOINT}/medic`, // TODO: Make this dynamic to handle multiple DBs
```
Comment (njuguna-n) [inline redis-worker/src/config.ts:12]: Created an issue to handle multiple DBs -> https://github.com/medic/cht-sync/issues/86

**Review by witash — COMMENTED** [review]
think we can remove .js from source control and add npx tsc to Dockerfile

looks really good, performance on local was about 1000 changes/second

I think it should be ok, but something to be aware of is we are adding a possibility for data to get lost; If logstash send the data to redis, and redis for whatever reason can't send data to postgres, logstash will advance the sequence and move on. Storing the redis queue in a persistent volume should be enough though.

I haven't yet tested what happens when logstash performance is better than postgrest/postgres, do you know? Eventually either redis will get 'full' somehow, or if no limits it will try put the entire changes feed in redis, which would use a ton of disk, although maybe that doesn't matter. if there are limits how does logstash respond?

File: redis-worker/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        console.log('No data in queue, waiting...');
+        await new Promise(resolve => setTimeout(resolve, 1000 * 10)); // 10 seconds
+        continue;
+      }
+
+      await updatePostgrest(data);
+      console.log('Data processed:', data);
+
+      const removedCount = await redisClient.lTrim(config.redis.key!, config.batchSize, -1);
+      console.log(`Updated ${data.length} items to PostgREST, removed ${removedCount} from queue.`);
```
Comment (witash) [inline redis-worker/src/index.ts:29]: its a small thing but ltrim returns OK not the count so this logs removed OK from queue

Comment (njuguna-n) [reply redis-worker/src/index.ts:29]: Yeah, I forgot to edit that log to remove that part. I will do that.

File: redis-worker/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  try {
+    while (true) {
+      const data = await redisClient.lRange(config.redis.key!, 0, config.batchSize - 1);
+      if (data.length === 0) {
+        console.log('No data in queue, waiting...');
+        await new Promise(resolve => setTimeout(resolve, 1000 * 10)); // 10 seconds
+        continue;
+      }
+
+      await updatePostgrest(data);
```
[the remark this answers is not in the document]
Comment (njuguna-n) [reply redis-worker/src/index.ts:25]: Good catch. I wiil make the change.

**Comment by witash** [comment]
answering my own question; redis list has a maximum of 2^32 elements, which it will never reach so I guess it will keep queueing until it runs out of disk space; if the logstash plugin can't connect to redis or can't lpush, it doesn't update the sequence and keeps retrying, so that's ok, probably we just need to manage the volume size on production deployments

this branch does depend on the persistent logstash sequence thing...as is, the redis queue is persistent but the logstash sequence isn't, so restarting often leads to a lot of duplicate elements in the queue; it doesn't lead to duplicate data but just has some kind of weird behavior

**Comment by njuguna-n** [comment]
When I was testing with the BRAC instance logstash was better than postgrest and like you said it led to the redis list becoming bigger and taking up more space. If there is a big discrepancy in the input and output throughputs logstash seems to rate limit itself and decrease the input throughput. Increasing the batch size being sent to postgrest can definitely help in managing this scenario.

**Comment by njuguna-n** [comment]
@lorerod do you mind helping me out with the failing test? All the containers seem to be created okay and the tests all pass locally so not sure what is going on in CI.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -23,6 +23,11 @@ jobs:
       COUCHDB_HOST: "couchdb"
       COUCHDB_PORT: 5984
       COUCHDB_SECURE: false
+      REDIS_HOST: redis
+      REDIS_PORT: 6379
+      REDIS_KEY: medic
+      POSTGREST_ENDPOINT: postgrest:3000
+      BATCH_SIZE: 100
```
Comment (lorerod) [inline .github/workflows/test.yml:30]:
```suggestion
      REDIS_BATCH_SIZE=100
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -23,6 +23,11 @@ jobs:
       COUCHDB_HOST: "couchdb"
       COUCHDB_PORT: 5984
       COUCHDB_SECURE: false
+      REDIS_HOST: redis
```
Comment (lorerod) [inline .github/workflows/test.yml:26]:
```suggestion
      POSTGRES_HOST=postgres 
      REDIS_HOST: redis
```

File: redis-worker/src/config.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+const config = {
+  redis: {
+    host: process.env.REDIS_HOST,
+    port: process.env.REDIS_PORT,
+    key: process.env.REDIS_KEY,
+  },
+  postgrest: {
+    url: `http://${process.env.POSTGREST_ENDPOINT}/medic`,
+  },
+  batchSize: parseInt(process.env.BATCH_SIZE ?? '100', 10),
```
Comment (lorerod) [inline redis-worker/src/config.ts:14]: Please double-check if this also needs to be REDIS_BATCH_SIZE.

**Review by lorerod — CHANGES_REQUESTED** [review]
@njuguna-n I reproduced the CI failure locally using the environment variables defined in `.github/workflows/test.yml`. I completed the environment variables and renamed one, and the test is now passing on my local machine. Please let me know if this works also for you.

**Comment by njuguna-n** [comment]
Thanks @lorerod. I have renamed the variable to reduce the likelihood of such errors in the future.

**Review by witash — APPROVED** [review]
LGTM

**Review by lorerod — APPROVED** [review]
Thank you, @njuguna-n. 
Just as a comment this repo definitely needs an enhancement to the integration tests (we need to rename e2e to integration). I have already added the [ticket](https://github.com/medic/cht-sync/issues/81) to my to-do list.

**Pull request merged**